server/src/main/java/org/opensearch/action/admin/indices/scale/searchonly/ScaleIndexShardSyncManager.java

```diff
@@ -0,0 +1,217 @@
+
+        for (Map.Entry<String, List<ShardId>> entry : nodeShardGroups.entrySet()) {
+            final String nodeId = entry.getKey();
+            final List<ShardId> shards = entry.getValue();
+            final DiscoveryNode targetNode = clusterService.state().nodes().get(nodeId);
+
+            if (targetNode == null) {
+                groupedListener.onFailure(new IllegalStateException("Node [" + nodeId + "] not found"));
+                continue;
+            }
+
+            sendNodeRequest(targetNode, index, shards, groupedListener);
+        }
+    }
+
+    /**
+     * Sends a sync request to a specific node for the given shards.
+     * <p>
+     * This method creates and sends a transport request to perform shard synchronization
+     * on a target node, registering appropriate response and error handlers.
+     *
+     * @param targetNode the node to send the request to
+     * @param index      the name of the index being scaled
+     * @param shards     the list of shards to synchronize on the target node
+     * @param listener   the listener to notify with the response
+     */
+    void sendNodeRequest(DiscoveryNode targetNode, String index, List<ShardId> shards, ActionListener<ScaleIndexNodeResponse> listener) {
+        transportService.sendRequest(
+            targetNode,
+            transportActionName,
+            new ScaleIndexNodeRequest(index, shards),
+            new TransportResponseHandler<ScaleIndexNodeResponse>() {
+                @Override
+                public ScaleIndexNodeResponse read(StreamInput in) throws IOException {
+                    return new ScaleIndexNodeResponse(in);
+                }
+
+                @Override
+                public void handleResponse(ScaleIndexNodeResponse response) {
+                    listener.onResponse(response);
+                }
+
+                @Override
+                public void handleException(TransportException exp) {
+                    listener.onFailure(exp);
+                }
+
+                @Override
+                public String executor() {
+                    return ThreadPool.Names.SAME;
+                }
+            }
+        );
+    }
+
+    /**
+     * Aggregates node responses and verifies that no shard reports uncommitted operations or a pending sync.
+     * <p>
+     * This validation ensures that all shards are in a consistent state before proceeding with
+     * a scale operation. If any shard reports conditions that would make scaling unsafe, the
+     * operation is failed with detailed information about which shards need more time.
+     *
+     * @param responses the collection of responses from all nodes
+     * @param listener  the listener to notify with the aggregated result
+     */
+    void validateNodeResponses(Collection<ScaleIndexNodeResponse> responses, ActionListener<ScaleIndexResponse> listener) {
```

Comment:
Is this method used in the production code? It looks like it's only called from tests.

Reply:
Yes Froh it is used.

In the `handleShardSyncResponses` method (part of `TransportScaleIndexAction`), the `validateNodeResponses`  is called to verify that all shards are in a consistent state before proceeding with the scale-down operation. 
The `validateNodeResponses` method performs some critical validations. It iterates through all shard responses from all nodes and checks if any shard has uncommitted operations and needs synchronization If any shards fail these checks, it collects their IDs and generates a detailed error message.

Reply:
Got it, thanks! The diff for `TransportScaleIndexAction` wasn't fetched, so I couldn't find where it was used. It makes sense.